database/app/src/main/java/com/google/firebase/referencecode/database/MainActivity.java

```diff
@@ -29,32 +30,35 @@
 public class MainActivity extends AppCompatActivity implements MainActivityInterface {
 
     private static final String TAG = "MainActivity";
+    DatabaseReference myRef;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
+        FirebaseDatabase database = FirebaseDatabase.getInstance();
+        myRef = database.getReference("message");
     }
 
     @Override
-    public void basicReadWrite() {
-        // [START write_message]
-        // Write a message to the database
-        FirebaseDatabase database = FirebaseDatabase.getInstance();
-        DatabaseReference myRef = database.getReference("message");
-
-        myRef.setValue("Hello, World!");
-        // [END write_message]
-
+    public void valueRead() {
         // [START read_message]
         // Read from the database
+        /*
+        ValueEventListener provides a DataSnapshot of a whole node on which it's set,
+        so every time the data on myRef changes, you get a snapshot of the entire myRef node.
+         */
         myRef.addValueEventListener(new ValueEventListener() {
             @Override
             public void onDataChange(DataSnapshot dataSnapshot) {
                 // This method is called once with the initial value and again
                 // whenever data at this location is updated.
-                String value = dataSnapshot.getValue(String.class);
-                Log.d(TAG, "Value is: " + value);
+                // Since we get a snapshot of the data we create an Iterable to traverse through it.
```

Comment:
Why did you make this change? We wrote "Hello World" to the reference so it's a String not an Iterable